lib/testing/src/equality/equality.dart

```diff
@@ -0,0 +1,229 @@
+ *      'world': ["world", "wor" + "ld"],
+ *      'three': [2, 1 + 1]
+ *     }, areEqualityGroups);
+ *
+ * This tests:
+ *
+ * * comparing each object against itself returns true
+ * * comparing each object against an instance of an incompatible class
+ *     returns false
+ * * comparing each pair of objects within the same equality group returns
+ *     true
+ * * comparing each pair of objects from different equality groups returns
+ *     false
+ * * the hash codes of any two equal objects are equal
+ *
+ *
+ * When a test fails, the error message labels the objects involved in
+ * the failed comparison as follows:
+ *
+ *      "`[group i, item j]`" refers to the jth item in the ith equality group,
+ *       where both equality groups and the items within equality groups are
+ *       numbered starting from 1.  When either a constructor argument or an
+ *       equal object is provided, that becomes group 1.
+ *
+ */
+const REPETITIONS = 3;
+const NO_MISMATCH_FOUND = 'therewasnomismatchfoundhere';
+const Matcher areEqualityGroups = const _EqualityGroupMatcher();
+
+class _EqualityGroupMatcher extends Matcher {
+  static const FAILURE_REASON = 'failureReason';
+  const _EqualityGroupMatcher();
+
+  @override
+  Description describe(Description description) =>
+      description.add('to be equality groups');
+
+  @override
+  bool matches(Map<String, List<Object>> item, Map matchState) {
+    try {
+      _verifyEqualityGroups(item, matchState);
+      return true;
+    } on MatchError catch (e) {
+      matchState[FAILURE_REASON] = e.toString();
+      return false;
+    }
+  }
+
+  Description describeMismatch(item, Description mismatchDescription,
+      Map matchState, bool verbose) => mismatchDescription
+        .add(" ${matchState[FAILURE_REASON]}");
+
+
+  void _verifyEqualityGroups(Map<String, List<Object>> equalityGroups,
+      Map matchState) {
+    if (equalityGroups == null) {
+      throw new MatchError('Equality Group must not be null');
+    }
+    var equalityGroupsCopy = {};
+    equalityGroups.keys.forEach((String groupName) {
+      if (groupName == null) {
+        throw new MatchError('Group name must not be null');
+      }
+      var group = equalityGroups[groupName];
+      if (group == null) {
+        throw new MatchError('Group must not be null');
+      }
+      equalityGroupsCopy[groupName] = new List.from(group);
+    });
+
+    // Run the test multiple times to ensure deterministic equals
+    for (var run in range(REPETITIONS)) {
+      _checkBasicIdentity(equalityGroupsCopy, matchState);
+      _checkGroupBasedEquality(equalityGroupsCopy);
+    }
+  }
+
+  void _checkBasicIdentity(Map<String, List<Object>> equalityGroups,
+      Map matchState) {
+    var flattened = equalityGroups.values.expand((group) => group);
+    for (var item in flattened) {
+      if (item == _NotAnInstance.EQUAL_TO_NOTHING) {
+        throw new MatchError("$item must not be Object#equals to an arbitrary "
```

Comment:
Replace uses of `Object#equals` with just "equal[s]" as the case warrants. It's verbose, and Dart Objects don't define an `equals` method, they define `operator ==` which is even more verbose.